support/cas-server-support-memcached-ticket-registry/build.gradle

```diff
@@ -1,19 +1,21 @@
 description = "Apereo CAS Memcached Integration"
 dependencies {
-    
+
     implementation libraries.memcached
     implementation libraries.kryo
 
     compile project(":core:cas-server-core-logging")
     compile project(":core:cas-server-core-tickets")
     compile project(":core:cas-server-core-services")
     compile project(":core:cas-server-core-configuration")
+    compile project(":support:cas-server-support-oauth")
```

Comment:
This line needs to go away. It simply forces everyone to run/enable oauth automatically, when they begin to use memcached. The test dependency too. 

In order to ensure oauth (and everything else) is accounted for, there will need to be a more dynamic strategy where each module on its own begins to "inject" relevant components into the kryo map, if kryo is found and enabled at runtime. Doing this statically causes unforeseen caveats.


All else looks fine. Thank you!

Reply:
In THAT case - I would much rather fix this by removing the "this.kryo.setAutoReset(false)" in KryoTranscoder.  That's the source of the problem, compounded by the "this.kryo.setRegistrationRequired(false)" so that missing registrations are found at Serialization time, rather than having a bad deserialization.  That WOULD mean a small increase in overhead for serializing and deserializing (both space and processing) an object containing an unregistered class.  But - I'd much rather that, than having a very fragile (IMO) approach as now, where programmers have to know that any class that could be serialized needs to be registered with Kryo, and there's no safety net when that doesn't happen.  Thoughts?

Reply:
I would support that approach; though rather than removing the logic it would be best to expose options for those two operations, such that if X = true, auto-registration is turned on and all else is skipped or otherwise, the existing approach is kept. This I believe should present the least number of surprising side-effects.

Reply:
Sorry - not following you in the details here.  If there's an option, what should the default be?  Under what conditions would the alternative be invoked?  From inside CAS or from someone overriding from a cas overlay project?

Reply:
> If there's an option, what should the default be?

Same thing as it is today. Auto-registration is off.

> Under what conditions would the alternative be invoked?

Given deployer's consent, the option would be tweaked. It's a decision one might have to make.

> From inside CAS or from someone overriding from a cas overlay project?

At the overlay project. The default value for the option preserves the behavior today, and YOU in the overlay would tweak the option to enable auto-registration. That solves the problem for you, and keeps the existing behavior for those who are happy with their existing configuration and have no issues.

Reply:
Thanks, that helps.  I'm still learning my way around the whole CAS code base, and also around Gradle which is new to me.  Is there anything comparable that already exists that I could take a look at?  e.g. some property that is exposed through the War Overlay, that is used inside the CAS code?

Reply:
Sure. There is `cas.server.name` mapped to `CasServerProperties.java`. Plenty of others.

If you review the master branch, in `BaseMemcachedProperties`, there are options for this sort of thing: `kryoAutoReset`, etc. They are ready and applied to the kryo component at initialization time.

Reply:
Well duh on me for the property question - of course adding a new property for memcached, accessible through MemcachedTicketRegistryProperties, is easy.  I just didn't make that connection for some reason :(

OK - here's my revised proposal to resolve this issue:

1) Change KryoTranscoder so that setRegistrationRequired is true instead of false.  That would work well with the AutoReset(false) - so that we detect any non-registered class at serialization time, rather than have problems at deserialization time.

2) Change KryoTranscoder so that it can have classes specified by name (e.g. "org.apereo.cas.ticket.accesstoken.OAuthAccessTokenExpirationPolicy"), and if Class.forName(xxx) doesn't throw a ClassNotFoundException, then register it with Kryo.  Use that approach for the three OAuth classes, and remove that Gradle dependency on cas-server-support-oauth.  So - that way - KryoTranscoder can still have the code to register all those classes - but will do so only if the user has included that OAuth dependency for their build.   (I thought of trying to access Kryo from inside those other modules so the registration could be local to the modules where the serializable classes are defined - but then that introduces a new dependency on the memcached module :()

3) Add a new property cas.ticket.registry.memcached.additionalSerializerClasses which takes a comma delimited list of fully qualified class names (e.g. "com.xxx.cas.MySerializableClass,com.xxx.cas.MySecondSerializableClass").  Use the same approach as with the other classes specified by name - i.e. if Class.forName doesn't throw a ClassNotFoundException, then register it with Kryo.  Perhaps on this one, throw a IllegalStateException ("No such class " + className) since the user should only use this config with classes that actually exist.  Better to throw the exception to help them debug, than to silently fail to register a class because they spelled the name wrong or something...

So - I think that addresses everything:
- More explicit failure of detecting an unregistered class at Serialization time.
- Add in the various potential CAS classes to the registered set, without introducing new cross-module dependencies.
- Allow for custom classes to be added to the Kryo registration.

Thoughts?

Reply:
1. See my earlier comment. As long as you expose an option and keep the current behavior, that's fine.
2. Negative. Class names may change, and forName() as performance side effects which should be avoided at all costs. As I mentioned, it's not just oauth. There are or may be plenty of other modules that might need the same, and hardcoding class names is not an acceptable approach. A module should be self contained, and all others with specialized serialization needs should take care of their own, on their own.
3. See #2.

Reply:
You see, the bottom-line motivation for the approach is quite simple: "it should just work". The expectation that folks would have to register class names via properties, and know how to do it, what the classes are and why, debug and troubleshoot to figure this stuff out is simple impractical. Nobody is ever going to understand it and by extension, nobody is ever going to do it. If you deploy OAuth or any other module, THAT MODULE already knows what things it might need to inject into the context already. Because it's doing weird things, and it knows best what needs to be serializable and how. The expectation that adopters would have to teach that to CAS is not reasonable, and dynamic behavior where the thing is then going to reflectively new up class names is also not that productive, because if for instance, oauth is already loaded on classpath at runtime, there is need to need to use reflections. Inject, fire and forget, which also keeps it consistent with everything else that CAS does dynamically, keeping modules separate with very little overhead (Consider how authn handlers are injected into the runtime).

Reply:
um.
1. It's BROKEN as it is.  Leaving it as is is not a valid option, IMO.  Why are you against that suggestion of throwing an exception if an unregistered class is encountered at Serialization time, given that the Deserialization is seriously broken right now?

2. What's the performance side effect?  This gets done once, at startup time.  It's not as though any performance hit would continue through the life of the running app.

I'm completely out of ideas on how to have the various modules handle their own registration, without introducing the cross-module dependencies that you want to avoid.  In fact - the current model doesn't follow this pattern.  The existing registration is for classes from other modules, but coded into the memcached module.

3.  See my #2.

Reply:
1. Correct. It's broken. For you. This is the first time this issue has been reported, and at least for me, I have deployed this module a number of times without ever having to deal with this issue. I don't know what's unique about your deployment, but I'd rather keep the existing behavior and expose a variation of it via an option to be on the safe side. If over time, the option proves to be more useful and indeed becomes da solution, we shall and by all means, switch.

2. A performance hit is a performance hit. If it can be avoid, it should be avoided. But more so, it's simply unnecessary to do for other reasons I outlined. There is no need for cross-module dependencies. This is how OIDC overrides OAuth, and this is how MongoDb can be used for GAuth and this is how you can register a JDBC authn handler into the authn engine without static linking. Memcached might not follow this approach just yet. It should, which is the root of my suggestion here.

I understand the complications up front, and will be more than happy to provide you with a foundation on master so you can resume work on here. Hopefully, that would serve as a good baseline. Short of that, simply live with the patch in your environment.

Reply:
By the way, in the interest of making progress I recommend the following: remove the oauth dependency for now so we can merge the PR. That would be very good. Once similar work is done on master, I'll ping this PR so you can review and implement a similar approach here as a new PR. Sound good?

Reply:
I'm good in the short term with removing the OAuth dependencies.  That will STILL end up fixing my explicit issue (the fact that the nested classes in MultiTimeUseOrTimeoutExpirationPolicy aren't registered with Kryo), without messing with the dependencies.  It just doesn't give any general ability to fix other, similar 'missing' classes from the registration.

Working on a more general approach in master, though, does sound good to me also.

Thanks!

Bob.

Reply:
Outstanding. Thanks. I'll work up a few examples and a put together a general framework for that sort of thing and we'll review together. Will ping the PR again when the changes are in.

Reply:
Hey - I just found that Kryo allows not only for an EXCEPTION to be thrown if an unregistered class is encountered, but also allows for a WARNING to be logged. What would you think about me adding:
         //  However, we would like at least a warning in the logs.
        this.kryo.setWarnUnregisteredClasses(true);

right after the:
          this.kryo.setRegistrationRequired(false);

That would at least help debug problems in the future, with the only noticeable change being that an additional warning is generated...

Reply:
Excellent find, and sure. Please do.